packages/frontend-v2/utils/plan/updatePlanOnDragEnd.ts

```diff
@@ -38,23 +47,52 @@ export const updatePlanOnDragEnd = (
       );
     }
 
-    if (!draggedCourse.data.current.isFromSidebar) {
-      // Course is from a term, so we need to move it, we don't need to move
-      // courses that are from the sidebar.
+    const draggedCourseDetails: ScheduleCourse2<unknown> =
+      draggedCourse.data.current.course;
 
-      // remove the class from the old term and add it to the new term
-      const oldTerm = scheduleTerms.find((term) =>
-        term.classes.some((course) => course.id === draggedCourse.id)
+    const oldTerm = scheduleTerms.find((term) =>
+      term.classes.some((course) => course.id === draggedCourse.id)
+    );
+
+    const isFromSidebar = draggedCourse.data.current.isFromSidebar;
+    const isSameTerm = !isFromSidebar && oldTerm && oldTerm.id === newTerm.id;
+
+    /*
+     * Prevent duplicate courses in the same term,
+     * don't need to display error if course is not changing terms.
+     */
+    if (
+      isCourseInTerm(
+        draggedCourseDetails.classId,
+        draggedCourseDetails.subject,
+        newTerm
+      ) &&
+      !isSameTerm
+    ) {
+      toast.error(
+        `Oops, ${getCourseDisplayString(
+          draggedCourseDetails
+        )} already exists in Year ${newTerm.year}, ${getSeasonDisplayWord(
+          newTerm.season
+        )}.`
       );
+      throw new Error("Duplicate course in term.");
```

Comment:
As a general question, should we be throwing errors for user-error that we've successfully handled?

Reply:
Currently it's just the way the updatePlanOnDragEnd API works where we throw errors to indicate "do not update the plan and make the API post request".

In most cases we don't even need to alert the user of the error though(dragging over a term not found etc.). So in those cases we just throw the error. In this case, we want to alert the user so we alert and throw.

A potential solution around this is as follows:
- Don't throw errors, and instead return them
- Return a flag indicating whether we want to alert the user of the error and then let the calling function take care of it

Not super sure if this is a huge improvement, we can discuss this and add this as a tech debt issue if there's benefit. But thanks for pointing it out!!